### PR TITLE
Remove allocation when a model with a supported SLA is being destroyed.

### PR DESCRIPTION
**Pull request by tasdomas**
## Description of change

This change is needed to fix the juju support for model SLAs.

## QA steps

Model creation and destruction should work as before.

## Documentation changes

Contact @cmars for questions regarding this feature.

### CODE REVIEW

File: cmd/juju/model/destroy.go [hunk 1]
```diff
@@ -216,6 +225,26 @@ func (c *destroyCommand) Run(ctx *cmd.Context) error {
...
+	}
+
+	bakeryClient, err := c.BakeryClient()
+	if err != nil {
+		ctx.Warningf("could not remove model allocation: %v", err)
+	}
+
+	budgetClient := getBudgetAPIClient(bakeryClient)
+
+	resp, err := budgetClient.DeleteAllocation(modelDetails.ModelUUID)
```
Comment (anastasiamac) [inline cmd/juju/model/destroy.go]: Making this call on CLI means it's not part of the transaction. We could be in situation where model destruction succeeded but allocation was not deleted, erred out for some reason.
Is it acceptable?

Comment (tasdomas) [inline cmd/juju/model/destroy.go]: Yes, this is fine.

File: cmd/juju/model/destroy.go [hunk 2]
```diff
@@ -216,6 +225,26 @@ func (c *destroyCommand) Run(ctx *cmd.Context) error {
 		return errors.Trace(err)
 	}
 
+	// Check if the model has an sla auth.
+	if !slaIsSet {
+		return nil
+	}
+
+	bakeryClient, err := c.BakeryClient()
```
Comment (cmars) [inline cmd/juju/model/destroy.go]: Consider moving lines 233-246 to an unexported method, `func (c *destroyCommand) deleteBudgetAllocation(modelDetails) error`. `deleteBudgetAllocation` can then return early on error, which would be logged here with a warning.

File: cmd/juju/model/destroy.go [hunk 3]
```diff
@@ -226,25 +226,32 @@ func (c *destroyCommand) Run(ctx *cmd.Context) error {
 	}
 
 	// Check if the model has an sla auth.
-	if !slaIsSet {
+	if slaIsSet {
+		err = c.removeModelAllocation(modelDetails.ModelUUID)
+		if err != nil {
+			ctx.Warningf("model allocation not removed: %v", err)
+		}
 		return nil
```
Comment (cmars) [inline cmd/juju/model/destroy.go]: No need to return early here.

File: cmd/juju/model/destroy.go [hunk 4]
```diff
@@ -216,6 +225,33 @@ func (c *destroyCommand) Run(ctx *cmd.Context) error {
 		return errors.Trace(err)
 	}
 
+	// Check if the model has an sla auth.
+	if slaIsSet {
+		err = c.removeModelAllocation(modelDetails.ModelUUID)
+		if err != nil {
+			ctx.Warningf("model allocation not removed: %v", err)
+		}
+		return nil
```
Comment (alesstimec) [inline cmd/juju/model/destroy.go]: dd

**Comment by tasdomas** [comment]
$$merge$$

**Pull request merged**